src/account/eth_account/eth_account.cairo

```diff
@@ -0,0 +1,256 @@
+        const INVALID_SIGNATURE: felt252 = 'EthAccount: invalid signature';
+        const INVALID_TX_VERSION: felt252 = 'EthAccount: invalid tx version';
+        const UNAUTHORIZED: felt252 = 'EthAccount: unauthorized';
+    }
+
+    #[embeddable_as(SRC6Impl)]
+    impl SRC6<
+        TContractState,
+        +HasComponent<TContractState>,
+        +SRC5Component::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of interface::ISRC6<ComponentState<TContractState>> {
+        /// Executes a list of calls from the account.
+        ///
+        /// Requirements:
+        ///
+        /// - The transaction version must be `TRANSACTION_VERSION` for actual transactions.
+        /// For simulations, the version must be `QUERY_VERSION`.
+        fn __execute__(
+            self: @ComponentState<TContractState>, mut calls: Array<Call>
+        ) -> Array<Span<felt252>> {
+            // Avoid calls from other contracts
+            // https://github.com/OpenZeppelin/cairo-contracts/issues/344
+            let sender = get_caller_address();
+            assert(sender.is_zero(), Errors::INVALID_CALLER);
+
+            // Check tx version
+            let tx_info = get_tx_info().unbox();
+            let version = tx_info.version;
+            if version != TRANSACTION_VERSION {
+                assert(version == QUERY_VERSION, Errors::INVALID_TX_VERSION);
+            }
+
+            execute_calls(calls)
+        }
+
+        /// Verifies the validity of the signature for the current transaction.
+        /// This function is used by the protocol to verify `invoke` transactions.
+        fn __validate__(self: @ComponentState<TContractState>, mut calls: Array<Call>) -> felt252 {
+            self.validate_transaction()
+        }
+
+        /// Verifies that the given signature is valid for the given hash.
+        fn is_valid_signature(
+            self: @ComponentState<TContractState>, hash: felt252, signature: Array<felt252>
+        ) -> felt252 {
+            if self._is_valid_signature(hash, signature.span()) {
+                starknet::VALIDATED
+            } else {
+                0
+            }
+        }
+    }
+
+    #[embeddable_as(DeclarerImpl)]
+    impl Declarer<
+        TContractState,
+        +HasComponent<TContractState>,
+        +SRC5Component::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of interface::IDeclarer<ComponentState<TContractState>> {
+        /// Verifies the validity of the signature for the current transaction.
+        /// This function is used by the protocol to verify `declare` transactions.
+        fn __validate_declare__(
+            self: @ComponentState<TContractState>, class_hash: felt252
+        ) -> felt252 {
+            self.validate_transaction()
+        }
+    }
+
+    #[embeddable_as(DeployableImpl)]
+    impl Deployable<
+        TContractState,
+        +HasComponent<TContractState>,
+        +SRC5Component::HasComponent<TContractState>,
```

Comment:
We cannot reuse these implementations even if there's a syntax for that, because the inner _is_valid_signature implementation is not the same. We could use some generic impl logic to abstract that, and we have discussed that issue, but having different types for the public key makes this probably not worth it (if we want to add a common type on top of both we would add extra cost by needing to serialize and deserialize for example).

Reply:
Ahh right. Good points. I agree, it probably isn't worth the trouble

Reply:
Using generic types won't lead to that, but keeping a common type like `Span<felt252>` would do, because you need to cast into the concrete type at some point (like what dispatchers do)